dali/pipeline/operators/detection/box_encoder.cc

```diff
@@ -0,0 +1,157 @@
+    CalculateIousForBox(ious_row, boxes[bbox_idx]);
+  }
+
+  return ious;
+}
+
+int BoxEncoder<CPUBackend>::FindBestBoxForAnchor(
+  int anchor_idx, const vector<float> &ious, int num_boxes) const {
+  int best_idx = 0;
+  float best_iou = ious[anchor_idx];
+
+  for (int bbox_idx = 1; bbox_idx < num_boxes; ++bbox_idx) {
+    if (ious[bbox_idx * NumAnchors() + anchor_idx] >= best_iou) {
+      best_iou = ious[bbox_idx * NumAnchors() + anchor_idx];
+      best_idx = bbox_idx;
+    }
+  }
+
+  return best_idx;
+}
+
+void BoxEncoder<CPUBackend>::MatchBoxesWithAnchors(
+  const vector<BoundingBox> &boxes, const int *labels, float *out_boxes, int *out_labels) const {
+  int num_boxes = boxes.size();
+  auto ious = CalculateIous(boxes);
+
+  for (int anchor_idx = 0; anchor_idx < NumAnchors(); ++anchor_idx) {
+    auto best_idx = FindBestBoxForAnchor(anchor_idx, ious, num_boxes);
+
+    // Filter matches by criteria
+    if (ious[best_idx * NumAnchors() + anchor_idx] > criteria_) {
+      WriteBoxToOutput(boxes[best_idx], out_boxes + anchor_idx * BoundingBox::kSize);
+      out_labels[anchor_idx] = labels[best_idx];
+    }
+  }
+}
+
+vector<BoundingBox> BoxEncoder<CPUBackend>::ReadBoxesFromInput(
+  const float *in_boxes, int num_boxes) const {
+  vector<BoundingBox> boxes;
+  boxes.reserve(num_boxes);
+  auto elem = in_boxes;
+
+  for (int idx = 0; idx < num_boxes; ++idx) {
+    boxes.push_back(BoundingBox::FromLtrb(elem, BoundingBox::NoBounds()));
+    elem += BoundingBox::kSize;
+  }
+
+  return boxes;
+}
+
+void BoxEncoder<CPUBackend>::WriteBoxToOutput(const BoundingBox &box, float *out_box_data) const {
+  auto out_box = box.AsCenterWh();
+
+  out_box_data[0] = out_box[0];
+  out_box_data[1] = out_box[1];
+  out_box_data[2] = out_box[2];
+  out_box_data[3] = out_box[3];
+}
+
+void BoxEncoder<CPUBackend>::WriteAnchorsToOutput(float *out_boxes, int *out_labels) const {
+  for (int idx = 0; idx < NumAnchors(); ++idx) {
+    WriteBoxToOutput(anchors_[idx], out_boxes + idx * BoundingBox::kSize);
+    out_labels[idx] = 0;
+  }
+}
+
+void BoxEncoder<CPUBackend>::RunImpl(SampleWorkspace *ws, const int idx) {
+  const auto &bboxes_input = ws->Input<CPUBackend>(0);
+  const auto &labels_input = ws->Input<CPUBackend>(1);
+
+  const auto num_boxes = bboxes_input.dim(0);
+
+  auto labels = labels_input.data<int>();
+  auto boxes = ReadBoxesFromInput(bboxes_input.data<float>(), num_boxes);
+
+  // Create output
+  auto bboxes_output = ws->Output<CPUBackend>(0);
+  bboxes_output->set_type(bboxes_input.type());
+  bboxes_output->Resize({NumAnchors(), BoundingBox::kSize});
+  auto out_boxes = bboxes_output->mutable_data<float>();
+
+  auto labels_output = ws->Output<CPUBackend>(1);
+  labels_output->set_type(labels_input.type());
+  labels_output->Resize({NumAnchors()});
+  auto out_labels = labels_output->mutable_data<int>();
+
+  WriteAnchorsToOutput(out_boxes, out_labels);
+  MatchBoxesWithAnchors(boxes, labels, out_boxes, out_labels);
```

Comment:
From my point of view, it is a bit confusing that something is going on after the last write. It prompts me to jump into the MatchBoxesFunction to really understand what's going on (since from RunImpl i dont get the whole picture), and suprrisingly there is a write call buried inside. I think that this function should return the prepared data, and in line 133 we should have the write call. This way it is easy to understand the flow of the op from RunImpl.

Reply:
Fixed